client/db/src/lib.rs

```diff
@@ -544,7 +544,7 @@ impl<Block: BlockT> BlockImportOperation<Block> {
 }
 
 impl<Block: BlockT> sc_client_api::backend::BlockImportOperation<Block> for BlockImportOperation<Block> {
-	type State = CachingState<RefTrackingState<Block>, Block>;
+	type State = SyncingCachingState<RefTrackingState<Block>, Block>;
```

Comment:
It would be really nice to have here `CachingState`, but that doesn't work and requires that we do some workarounds (`disable_syncing`)

Reply:
`into_cache_changes` leaves the local cache empty. So `drop` would have nothing to sync anyway, and it would not have to take any locks in such case. You could probably remove the `disable_syncing` workaround and simply do nothing if the cache is empty.

Reply:
I had dead locks when I did not disable syncing for the state in block import. Did not really track it down. I think we should get rid of the `get_import_lock` function in the long term and than we should be able to maybe remove `disable_syncing`.

Reply:
The deadlock is because in `drop` you don't check if the cache is empty before taking the lock. If add that check first it should be fine, as you'd always empty the cache on import.
I agree that `get_import_lock` should be removed eventually.

Reply:
On the other hand `disable_syncing` make the code more explicit and easier to follow